src/main/java/com/cloudbees/jenkins/plugins/bitbucket/endpoints/AbstractBitbucketEndpoint.java

```diff
@@ -0,0 +1,139 @@
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package com.cloudbees.jenkins.plugins.bitbucket.endpoints;
+
+import com.cloudbees.plugins.credentials.CredentialsMatchers;
+import com.cloudbees.plugins.credentials.CredentialsProvider;
+import com.cloudbees.plugins.credentials.common.StandardUsernamePasswordCredentials;
+import com.cloudbees.plugins.credentials.domains.URIRequirementBuilder;
+import edu.umd.cs.findbugs.annotations.CheckForNull;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import hudson.model.AbstractDescribableImpl;
+import hudson.security.ACL;
+import jenkins.model.Jenkins;
+import org.apache.commons.lang.StringUtils;
+
+/**
+ * Represents a {@link BitbucketCloudEndpoint} or a {@link BitbucketServerEndpoint}.
+ *
+ * @since 2.2.0
+ */
+public abstract class AbstractBitbucketEndpoint extends AbstractDescribableImpl<AbstractBitbucketEndpoint> {
+
+    /**
+     * {@code true} if and only if Jenkins is supposed to auto-manage hooks for this end-point.
+     */
+    private final boolean manageHooks;
+
+    /**
+     * The {@link StandardUsernamePasswordCredentials#getId()} of the credentials to use for auto-management of hooks.
+     */
+    @CheckForNull
+    private final String credentialsId;
+
+    /**
+     * Constructor.
+     *
+     * @param manageHooks   {@code true} if and only if Jenkins is supposed to auto-manage hooks for this end-point.
+     * @param credentialsId The {@link StandardUsernamePasswordCredentials#getId()} of the credentials to use for
+     *                      auto-management of hooks.
+     */
+    AbstractBitbucketEndpoint(boolean manageHooks, @CheckForNull String credentialsId) {
+        this.manageHooks = manageHooks && StringUtils.isNotBlank(credentialsId);
+        this.credentialsId = manageHooks ? credentialsId : null;
+    }
+
+    /**
+     * Optional name to use to describe the end-point.
+     *
+     * @return the name to use for the end-point
+     */
+    @CheckForNull
+    public abstract String getDisplayName();
```

Comment:
Which is why the name is "${displayName == null ? serverUrl : displayName + ' ('+serverUrl+')'}"

If we don't display the server url then we end up with people putting the server url in the display name, or worse putting the wrong server url in the display name.

If we provide the display name then we need to dedup on *both* otherwise the display name is meaningless or requires the server url anyway.

For the user to be able to trust the server url part of the display name, it has to be appended programmatically not relying on the admin.